raft/src/main/java/org/apache/kafka/raft/internals/SerdeRecordsIterator.java

```diff
@@ -0,0 +1,231 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft.internals;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import org.apache.kafka.common.protocol.DataInputStreamReadable;
+import org.apache.kafka.common.protocol.Readable;
+import org.apache.kafka.common.record.DefaultRecordBatch;
+import org.apache.kafka.common.record.FileRecords;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MutableRecordBatch;
+import org.apache.kafka.common.record.Records;
+import org.apache.kafka.common.utils.BufferSupplier;
+import org.apache.kafka.raft.BatchReader.Batch;
+import org.apache.kafka.raft.RecordSerde;
+
+public final class SerdeRecordsIterator<T> implements Iterator<Batch<T>>, AutoCloseable {
+    private final Records records;
+    private final RecordSerde<T> serde;
+    private final BufferSupplier bufferSupplier;
+    private final int maxBatchSize;
+
+    private Optional<Iterator<MutableRecordBatch>> nextBatches = Optional.empty();
+    private Optional<Batch<T>> nextBatch = Optional.empty();
+    // Buffer used to as the backing store for nextBatches if needed
+    private Optional<ByteBuffer> allocatedBuffer = Optional.empty();
+    // Number of bytes from records that read
+    private int bytesRead = 0;
+    private boolean isClosed = false;
+
+    public SerdeRecordsIterator(
+        Records records,
+        RecordSerde<T> serde,
+        BufferSupplier bufferSupplier,
+        int maxBatchSize
+    ) {
+        this.records = records;
+        this.serde = serde;
+        this.bufferSupplier = bufferSupplier;
+        this.maxBatchSize = maxBatchSize;
+    }
+
+    @Override
+    public boolean hasNext() {
+        checkIfClosed();
+
+        if (!nextBatch.isPresent()) {
+            nextBatch = nextBatch();
+        }
+
+        return nextBatch.isPresent();
+    }
+
+    @Override
+    public Batch<T> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException("Batch iterator doesn't have any more elements");
+        }
+
+        Batch<T> batch = nextBatch.get();
+        nextBatch = Optional.empty();
+
+        return batch;
+    }
+
+    @Override
+    public void close() {
+        isClosed = true;
+        allocatedBuffer.ifPresent(bufferSupplier::release);
+        allocatedBuffer = Optional.empty();
+    }
+
+    private void checkIfClosed() {
+        if (isClosed) {
+            throw new IllegalStateException("Serde record batch itererator was closed");
+        }
+    }
+
+    private Optional<Iterator<MutableRecordBatch>> nextBatches() {
```

Comment:
Is the Optional necessary here? Could we instead return an empty Iterator?

Reply:
Yes. Done.